### PR TITLE
[21.05] Pl 127635 refactor fc ceph

### PR DESCRIPTION
**Pull request by ctheune**
@flyingcircusio/release-managers

## Release process

Impact:

n/a

Changelog:

(internal ceph utility refactoring, no changelog entry needed?)

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)

Ensure proper argument handling and data structure handling when calling external programs. We refactored things to unify the different ways of calling external programs and at the same time fold various methods of interacting with CLI and JSON CLI programs into a homogenous internal API.

- [x] Security requirements tested? (EVIDENCE)

Pure internal refactoring, unit tests and hydra tests still work.

### CODE REVIEW

File: pkgs/fc/ceph/fc/ceph/util.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+class JSONRunner(object):
+    def __init__(self, runner):
+        self.runner = runner
+
+    def __run__(self, name, *args, **kw):
+        result = getattr(self.runner, name)(*args, **kw)
+        result = json.loads(result)
+        return result
+
```
Comment (osnyx) [inline pkgs/fc/ceph/fc/ceph/util.py:26]: For an easier superficial code structure, I suggest marking the tool-specific overrides with a comment to make the role of these member functions obvious at first glance. 

```suggestion

# tool-specific overrides of JSON formatting and extraction

```

File: pkgs/fc/ceph/fc/ceph/manage.py [hunk 2]
```diff
@@ -420,18 +344,13 @@ def deactivate(self, flush=True):
...
-                    "--osd-data",
-                    self.datadir,
-                    "--osd-journal",
-                    self._locate_journal_lv(),
-                ],
-                check=True,
+            run.ceph_osd(
+                # fmt: off
+                "-i", str(self.id),
+                "--flush-journal", "--osd-data",
```
Comment (osnyx) [inline pkgs/fc/ceph/fc/ceph/manage.py]: The grouping of arguments is a bit odd here. I'd expect argument name and argument value to be on the same line (e.g `"--osd-data", self.datadir,`).

**Review by osnyx — APPROVED** [review]

File: pkgs/fc/ceph/fc/ceph/util.py (hunk 1 above)
Comment (ctheune) [reply pkgs/fc/ceph/fc/ceph/util.py:26]: thanks. added more docs as suggested.

File: pkgs/fc/ceph/fc/ceph/manage.py (hunk 2 above)
Comment (ctheune) [reply pkgs/fc/ceph/fc/ceph/manage.py]: good catch!

File: pkgs/fc/ceph/fc/ceph/util.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    def __init__(self, runner):
+        self.runner = runner
+
+    def __run__(self, name, *args, **kw):
+        result = getattr(self.runner, name)(*args, **kw)
+        result = json.loads(result)
+        return result
+
+    # Tool-specific overrides to ensure that the invoed tools
```
Comment (osnyx) [inline pkgs/fc/ceph/fc/ceph/util.py]:
```suggestion
    # Tool-specific overrides to ensure that the involved tools
```

typo-alarm 🚨

**Review by osnyx — CHANGES_REQUESTED** [review]

**Review by osnyx — APPROVED** [review]

**Pull request merged**